### ISSUE TITLE
Issue with special character (e.g. '/') when renaming

### ISSUE THREAD

**Issue by nguillaumin** [jangler/taffy#2]
Hi,

Thanks for this great, yet simple tool.

I'm having issues when renaming files where the tags contains special characters, especially slashes `/`:

```bash
$ taffy --rename "%n - %T" */*.ogg

/usr/lib/ruby/gems/2.4.0/gems/taffy-1.2.1/bin/taffy:177:in `rename': No such file or directory @ rb_file_s_rename - (Fight-War_of_Words/12 - Reality, a New Beginning - Jesus Saves.ogg, Fight-War_of_Words/12 - Reality, a New Beginning / Jesus Saves.ogg) (Errno::ENOENT)
        from /usr/lib/ruby/gems/2.4.0/gems/taffy-1.2.1/bin/taffy:177:in `block (2 levels) in <top (required)>'
        from /usr/lib/ruby/gems/2.4.0/gems/taglib-ruby-0.7.1/lib/taglib/base.rb:8:in `open'
        from /usr/lib/ruby/gems/2.4.0/gems/taffy-1.2.1/bin/taffy:150:in `block in <top (required)>'
        from /usr/lib/ruby/gems/2.4.0/gems/taffy-1.2.1/bin/taffy:149:in `each'
        from /usr/lib/ruby/gems/2.4.0/gems/taffy-1.2.1/bin/taffy:149:in `<top (required)>'
        from /usr/bin/taffy:22:in `load'
        from /usr/bin/taffy:22:in `<main>'
```

The issue is that the song title is `Reality, a New Beginning / Jesus Saves` which contains a slash, which is a problematic character to have in a filename even under Linux (which I'm using).

I guess a naive approach could be to just skip it, or replace it with something like `-`.

**Comment by nguillaumin** [comment]
As a workaround, I added around l. 71:

```ruby
name = name.gsub('/', '-')
```

**Comment by nguillaumin** [comment]
There may be other special character on other OSes, like `\`, `:` on Windows...

**Comment by nguillaumin** [comment]
Confirmed `:` is a problem too, when I tried to copy a renamed file from my Linux box to a FAT formatted USB drive.

What would be nice would be to have different rules to convert special chars. For example:
* `/` would perhaps make sense to convert to `-`: `Reality, a New Beginning / Jesus Saves` → `Reality, a New Beginning - Jesus Saves`
* `:` should be handled differently, for exampe `Le Carnaval des animaux: II. Poules et Coqs` it's not preceded by a space so naively converting it to `-` would result in `Le Carnaval des animaux- II. Poules et Coqs`

I also suspect making it an option would be ideal, since different users would probably have 
 different opinions...

**Comment by jangler** [comment]
Oh weird, I thought it did this already. OK, will look into it.

**Comment by jangler** [comment]
Oh, have you tried `%_n - %_T` instead of `%n - %T`? That should replace all the special characters with underscores. I don't think there's currently any way to remove them altogether, though.

**Comment by nguillaumin** [comment]
Ah, fair enough I missed this option in the help message. However:

```bash
$ taffy --rename "%n - %_T" *.ogg
/usr/lib/ruby/gems/2.4.0/gems/taffy-1.2.1/bin/taffy:177:in `rename': No such file or directory @ rb_file_s_rename - (12 - Reality, a New Beginning - Jesus Saves.ogg, ./12 - Reality,_a_New_Beginning_/_Jesus_Saves.ogg) (Errno::ENOENT)
        from /usr/lib/ruby/gems/2.4.0/gems/taffy-1.2.1/bin/taffy:177:in `block (2 levels) in <top (required)>'
        from /usr/lib/ruby/gems/2.4.0/gems/taglib-ruby-0.7.1/lib/taglib/base.rb:8:in `open'
        from /usr/lib/ruby/gems/2.4.0/gems/taffy-1.2.1/bin/taffy:150:in `block in <top (required)>'
        from /usr/lib/ruby/gems/2.4.0/gems/taffy-1.2.1/bin/taffy:149:in `each'
        from /usr/lib/ruby/gems/2.4.0/gems/taffy-1.2.1/bin/taffy:149:in `<top (required)>'
        from /usr/bin/taffy:22:in `load'
        from /usr/bin/taffy:22:in `<main>'

```

It looks like `/` is still not covered (Moreover I'd actually want to preserve the spaces and not replace them by underscores I guess).

**Comment by jangler** [comment]
Hm! Alright, well the forward slash at the very least needs to be covered. IIRC the code just uses Ruby's built-in shell escape functionality to determine which characters are special, so it looks like it'll need to be replaced with some custom logic.

**Comment by nguillaumin** [comment]
Thanks, finally got a chance to test this and it works perfectly.

**Comment by nguillaumin** [comment]
I just found out that there are a few others that need to be covered. I got an issue with a `?` and `"`, but looking at the [Microsoft docs](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365247(v=vs.85).aspx) it seems all the reserved characters are:

> * < (less than)
> * \> (greater than)
> * : (colon)
> * " (double quote)
> * / (forward slash)
> * \ (backslash)
> * | (vertical bar or pipe)
> * ? (question mark)
> * \* (asterisk)

Should I open a separate issue?

**Comment by jangler** [comment]
I'll just reopen this one. I'm surprised that `"` isn't covered by Ruby's shell escape.

**Comment by jangler** [comment]
Code was stupid, should be better now. 😛

**Comment by nguillaumin** [comment]
Thanks, sorry I didn't get time to test this earlier, but with the latest version all the spaces get stripped. For example if the song title is "Here It Comes Slowly", I'm now getting `02 - HereItComesSlowly.ogg`. It also seems to strip non ASCII characters like accented letters, i.e. "Les Écorchés" becomes `06 - Lescorchs.ogg`

**Comment by jangler** [comment]
Space-stripping is intentional and should have been happening before, since spaces are "special characters" in the sense that they are annoying to deal with in the command line. If you want to preserve spaces, then you can do something like `--rename '% R - % T`, although this will replace other special characters with spaces as well.

I guess Ruby's regexes don't consider é and other Latin-1 type characters to be "word characters", which is atavistic in my opinion… not sure how to deal with that other than by adding a dependency on a library that actually knows Unicode character classes, or including a giant table.

**Comment by nguillaumin** [comment]
Thanks, for the spaces that's fine, I don't mind changing my command line invocation.

For the non-ASCII characters, well... That makes Taffy mostly useless to me with that change :wink: 

To me `shellescape` is not necessarily suitable to escape things here, because it's not about escaping a command to run through the shell, it's about escaping filenames to store on a file system. They have wildly different constraint, for example while spaces are indeed problematic for shell commands, they're perfectly fine on most file systems. `:` are fine to run through a shell, but they can't be used in a Windows file name. So it feels like we're using the wrong escape here?

So perhaps it just needs another approach, and have specific code to deal with the idiosyncrasies of various file systems? I came across a related article on this recently: http://beets.io/blog/paths.html

**Comment by jangler** [comment]
For some reason I actually use this ugly-as-hell bash function for sorting & naming my music files instead of Taffy, which explains why I haven't noticed these problems:

```bash
gettag () {
	echo "$1" | grep "^    $2 " | sed 's/^.\+: //' | \
		tr '[:upper:]' '[:lower:]' | tr -d "'" | \
		tr -s '`~!@#$%^&*()=+[{]}\|;:",<.>/? ' '_'
}
```

Which reminds me that Taffy doesn't need to know about every valid Unicode character, just the ones that aren't allowed in filenames. This logic should be easy to recreate in Ruby (obviously generalized to accomodate replacement characters and styles aside from the exact one I'm using here); I just need to make sure I'm not forgetting anything else stupid…

**Comment by jangler** [comment]
Forgot about this, making a note to see if I can get it done this weekend.

**Comment by jangler** [comment]
Do you want to check out the [special-characters](https://github.com/jangler/taffy/tree/special-characters) branch before I merge it with master? If that's convenient for you. I want to make sure I get it right this time. 😛

**Comment by nguillaumin** [comment]
Thanks, I gave it a shot and it's better, but I think we're still stripping too much. For example `'` is very common in song titles (I have 1115 files with a `'` in them), `&` too:

`06 - Nobody Knows You When You're Down & Out.ogg` → `06 - Nobody Knows You When Youre Down Out.ogg`.

Same with `!`, `.` or `-` or even `()`:

`10 - Being for the Benefit of Mr. Kite! - I Want You (She’s So Heavy) - Helter Skelter.ogg` → `10 - Being for the Benefit of Mr Kite I Want You She’s So Heavy Helter Skelter.ogg`

(Extreme example eh? I was surprised to find it, but [it's real](https://musicbrainz.org/recording/6ff31a54-b989-42e3-bcc9-cd06ce83edd1))

I would suggest only stripping characters that are actually problematic on file systems, not the shell, as per my previous list:

> * < (less than)
> * \> (greater than)
> * : (colon)
> * " (double quote)
> * / (forward slash)
> * \ (backslash)
> * | (vertical bar or pipe)
> * ? (question mark)
> * \* (asterisk)

**Comment by jangler** [comment]
That conflicts with my intended use of `--rename`, which is to make valid filenames which can also be used safely in the shell. Taffy does actually makes an exception for hyphens, though, since they're so common and one of the more obvious choices for a substitution character, so your Beatles example wouldn't be as butchered as you say. `!@^+[]{},.` are pretty innocuous and could be left alone, though.

**Comment by nguillaumin** [comment]
Hm, I'm confused about the use case I think: Shell scripts should be able to deal with any characters in filenames, as long as they're properly escaped or quoted? For example there are tons of shell scripts that can deal with filenames with spaces in them, or dashes? If that weren't the case a lot of things wouldn't work, and file names would be extremely restricted. Perhaps a better approach is to update the shell scripts you use to work with all kinds of file names? Can  you share the shell scripts you use with music files?

Alternatively, perhaps these are actually 2 use cases, and Taffy could have 2 flags: One `--rename-shell` to rename files for shell use, and `--rename` to just rename the files as-is?

**Comment by nguillaumin** [comment]
For example, here's a shell script I run over my music folder. It works fine with filenames that have spaces, dashes or other supposedly "unsafe" characters:

```bash
#!/bin/bash

shopt -s globstar

for i in **/*.ogg; do
    if ! oggz-validate "$i"; then
        echo $i
    fi
    RATE=$(ogginfo "$i" | grep ^Rate | cut -d ' ' -f 2)
    if [ -z $RATE ]; then
        echo $i: Missing rate
    elif [ ! $RATE -eq 44100 ]; then
        echo $i: Invalid rate $RATE
    fi
done
```

**Comment by jangler** [comment]
It's really not about scripts, which as you say should be smart enough to deal with arbitrary filenames; it's about little things like having to manually escape characters in the shell, `ls` quoting filenames with special characters in them, stuff like ctrl+left/right working differently on arguments with escaped spaces.

It seems like there are two use cases, but `--rename` has always been implemented in terms of filenames that are nice to use in the shell, so if anything, I feel like a less strict rename function should be called something else—but I can't think of anything that's as obvious as `--rename` and `--rename-shell`.

**Comment by nguillaumin** [comment]
Right, that kind of makes sense, although personally I'd rather have "proper" filenames and minor shell inconvenience than the opposite :stuck_out_tongue: Everyone's different I guess!

I'd be fine with keeping `--rename` as is and introducing a new one. Perhaps `--rename-unsafe` or `--rename-noshell`? `--rename-filesystem` / `--rename-fs`, `--rename-noescape`?

**Comment by jangler** [comment]
OK, I added a `--rename-fs` option. Good now?

**Comment by nguillaumin** [comment]
Yep, looks all good! Thanks!

**Comment by jangler** [comment]
Merged & pushed in version 1.4.0.
